pkg/querier/cardinality/request.go

```diff
@@ -0,0 +1,225 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package cardinality
+
+import (
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql/parser"
+	"golang.org/x/exp/slices"
+)
+
+const (
+	RequestTypeLabelNames  = RequestType(iota)
+	RequestTypeLabelValues = RequestType(iota)
+
+	minLimit     = 0
+	maxLimit     = 500
+	defaultLimit = 20
+
+	stringSeparator = '|'
+)
+
+type RequestType int
+
+type LabelNamesRequest struct {
+	Matchers []*labels.Matcher
+	Limit    int
+}
+
+// Strings returns a full representation of the request. This returned string can be
+// used to uniquely identify the request.
+func (r *LabelNamesRequest) String() string {
+	b := strings.Builder{}
+
+	// Add matchers.
+	for idx, matcher := range r.Matchers {
+		if idx > 0 {
+			b.WriteRune(',')
+		}
+		b.WriteString(matcher.String())
+	}
+
+	// Add limit.
+	b.WriteRune(stringSeparator)
+	b.WriteString(strconv.Itoa(r.Limit))
+
+	return b.String()
+}
+
+func (r *LabelNamesRequest) RequestType() RequestType {
+	return RequestTypeLabelNames
+}
+
+// TODO doc
+// TODO test (GET and POST)
+func DecodeLabelNamesRequest(r *http.Request) (*LabelNamesRequest, error) {
+	var (
+		parsed = &LabelNamesRequest{}
+		err    error
+	)
+
+	err = r.ParseForm()
+	if err != nil {
+		return nil, err
+	}
+
+	parsed.Matchers, err = extractSelector(r)
+	if err != nil {
+		return nil, err
+	}
+
+	parsed.Limit, err = extractLimit(r)
+	if err != nil {
+		return nil, err
+	}
+
+	return parsed, nil
+}
+
+type LabelValuesRequest struct {
+	LabelNames []model.LabelName
+	Matchers   []*labels.Matcher
+	Limit      int
+}
+
+// Strings returns a full representation of the request. This returned string can be
+// used to uniquely identify the request.
+func (r *LabelValuesRequest) String() string {
+	b := strings.Builder{}
+
+	// Add label names.
+	for idx, name := range r.LabelNames {
+		if idx > 0 {
+			b.WriteRune(',')
+		}
+		b.WriteString(string(name))
+	}
+
+	// Add matchers.
+	b.WriteRune(stringSeparator)
+	for idx, matcher := range r.Matchers {
+		if idx > 0 {
+			b.WriteRune(',')
+		}
+		b.WriteString(matcher.String())
+	}
+
+	// Add limit.
+	b.WriteRune(stringSeparator)
+	b.WriteString(strconv.Itoa(r.Limit))
+
+	return b.String()
+}
+
+func (r *LabelValuesRequest) RequestType() RequestType {
+	return RequestTypeLabelValues
+}
+
+// DecodeLabelValuesRequest parses query params from GET requests and parses request body from POST requests.
+// TODO test (GET and POST)
+func DecodeLabelValuesRequest(r *http.Request) (*LabelValuesRequest, error) {
```

Comment:
hint: if `DecodeLabelValuesRequest` works with `r.Form` only (type `url.Values`), you don't need to test both GET and POST here :)

Practically, testing with both GET/POST tests `r.ParseForm()` more than actual parameter parsing.

Reply:
Well I just wanted to ensure we use `ParseForm()`. Unfortunately I read this comment too late, and I've already added that test (but it's a very small one).